### ISSUE TITLE
[vm/ffi] Support FFI on SIMARM(64)

### ISSUE THREAD

**Issue by sjindel-google** [dart-lang/sdk#37299]
This could be useful for testing. We would probably need to re-use some of the design of the DBC implementation.

**Comment by dcharkes** [comment]
WIll not implement.

**Comment by dcharkes** [comment]
cc @cskau-g

**Comment by eseidel** [comment]
Pulling on this thread today.  Looks like the arm simulator knows how to call out to "dart natives" via RedirectExternalReference:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/simulator_arm64.cc#L922
Which is inserted via:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/runtime_entry_arm64.cc#L36
and
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/native_entry.cc#L114
Which then inserts a special instruction which the simulator catches and does the redirect:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/simulator_arm64.cc#L1779
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/simulator_arm64.cc#L1690
I've not looked into the FFI code enough yet, but I suspect the code ends up similar.  An additional question with FFI may be translating calling conventions across architectures?  Not sure.  (In my case I'm currently running the ARM64 Simulator on ARM64, by forcing USING_SIMULATOR on, so I can likely skip any calling convention translation to get my hacks working.)

**Comment by eseidel** [comment]
Yeah, so this appears to be how natives get turned into exception instructions and thus turned into thunks out into the real thing:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/compiler/stub_code_compiler_arm64.cc#L869
Will go read and figure out how FFI does similar.

**Comment by eseidel** [comment]
My naive understanding:

NativeCallInstr is generated from the FlowGraphBuilder.
Importantly, when it's code is generated:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/compiler/backend/il_arm64.cc#L1274
It hops through StubCode::Call*Native() which were populated by the StubCompiler:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/compiler/stub_code_compiler_arm64.cc#L862
to bounce through *_native_wrapper_entry_point_
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/thread.h#L242
which were populated by native_entry to add the simulator escape instruction:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/native_entry.cc#L114

FfiCallInstr also of course comes from FlowGraphBuilder.
When it's code is generated:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/compiler/backend/il_arm64.cc#L1331
It makes no attempt to hop through a stub, so there is no hook for the Simulator to insert its redirection instruction:
https://github.com/dart-lang/sdk/blob/cc8b3f792eb245cfdcc7c6dce76f2eaf18061d18/runtime/vm/constants_arm64.h#L1332

Presumably the "fix" here would be to insert some sort of stub call (at least on the simulator) to let the simulator hook its redirection instruction.  Will read more.

**Comment by dcharkes** [comment]
It's been a while since I spend some thought on this issue. Let me see what I remember:

An important difference between the old native calls and the FFI is that the old native calls all use exactly the same signature:

https://github.com/dart-lang/sdk/blob/e1f153f7a2666ee20bc84081d8b288e99f561158/runtime/vm/native_entry.h#L42-L43

These happen to be just 3 normal registers and the normal return register for all simulated architectures. So the stub can be simply doing that.

For FFI calls, we can have arguments on the stack, and all argument registers and double argument registers. So the stub would need to work with a generic call (kind of how the callback trampolines work).

A second challenge is that our simulators only implement the subset of of instructions that the Dart compiler emits. That means we can't just load an arbitrary dll for the simulated architecture and simulate it. Instead, we should `DynamicLibrary.open` dlls from _host_ architecture. (This is kind of similar to the natives, because the simulator runs on the host architecture.) So the VM implementation for dlopen etc need to take that into account.

A third challenge has to do with how the code is setup. Currently, the computation of the native calling convention can only be done for the `TARGET_OS` and `TARGET_ARCH` due to ifdefs. When calling with an FFI function into a HOST function, we need to read from the registers/stack from the TARGET calling convention in the simulator and put the values in the HOST registers/stack.

The code is not setup to deal with this. Only the `TARGET` is available, everything is ifdeffed out.
In order to fix that the constants need to be rename-spaced. Undoing: https://dart-review.googlesource.com/c/sdk/+/134298 Second, the runtime/vm/compiler/ffi/native_calling_convention.h and its imports need to be templated in either host or target. Undoing: https://dart-review.googlesource.com/c/sdk/+/134297

So, this is a non-trivial undertaking.

@eseidel Why are you running a simulator on the same architecture as the host? What's your use case?

**Comment by dcharkes** [comment]
> The first demos I'm getting going involve running 100% of the dart code in the simulator, and to unblock that, I need Flutter to boot in the simulator, which requires fixing FFI (at last for HOST=TARGET workloads, specifically Arm64). 

Interesting!

Yes, HOST_ARCH==TARGET_ARCH with HOST_OS==TARGET_OS should make things a lot easier.

> The hack I'm trying today is I'm going to invent a new simulator break instruction, ala:
> 
> https://github.com/dart-lang/sdk/blob/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/simulator_arm64.cc#L1902
> 
> 
> Except I'll use something which doesn't actually do anything on a physical cpu, like `mov r1, r1` or maybe `b 1` but that Dart would never generate.
> Have FfiCallInstr generate said break right before the branch:
> 
> https://github.com/dart-lang/sdk/blob/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/compiler/backend/il_arm64.cc#L1410

Sounds like a good plan. Note that there's more than one branching point.
https://github.com/dart-lang/sdk/blob/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/compiler/backend/il_arm64.cc#L1446

> And then from within said break I will:
> 
> 1. decode the next instruction (get the branch I was supposed to take)
> 2. "restore" the simulator's cpu registers into the real cpu (not sure the exact set?  Possibly just all of them to start?)

These are the argument registers.

https://github.com/dart-lang/sdk/blob/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/constants_arm64.cc#L32-L38

The registers will only be enough if we don't actually need the stack for arguments.
https://github.com/dart-lang/sdk/blob/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/compiler/ffi/native_calling_convention.cc
You can see some of the calling convention choices be spelled out in
https://github.com/dart-lang/sdk/blob/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/compiler/ffi/native_calling_convention_test.cc
with
https://github.com/dart-lang/sdk/tree/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/compiler/ffi/unit_tests
And you can play around with the calling conventions on a compiler here: https://godbolt.org/

> 3. take the actual blr as it would have before, except now returning into this break stub.
> 4. "save" the real cpu's registers back into the simulator

These are the return registers:

https://github.com/dart-lang/sdk/blob/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/constants_arm64.h#L599-L601

PointerToMemory might actually work here as well. E.g. the caller reserves memory to in which a struct return value stored by the callee. That pointer is passed in in one of the argument registers. On arm64 you also need to pass R8 from simulator to host for that.

https://github.com/dart-lang/sdk/blob/6b35aa94b996863a3c1e4d345ea6e8756c7569b2/runtime/vm/compiler/ffi/native_calling_convention.cc#L666

> 5. resume simulating.
> 
> I think it might work? We'll hopefully know in a few hours.
> 
> I obviously have many challenges to overcome after that, but I think my reading yesterday (and your helpful reply) have a path to at least get something going and thus unblock running Flutter code within the Dart ARM simulator on ARM64.

One sanity check you could add in your prototype is that all `NativeLocation`s in your `NativeCallingConvention` are only the `NativeRegisterLocation`s that you implement in your first prototype. That way you can crash during compilation of an FFIcall rather than during execution.

I think registers only might not be enough and that you will have to do the stack as well for arguments. The `dart:ui` migration from old natives to `@Native`s (which use FFI) saw function signatures with up to 20 parameters. (But with the sanity check we'll figure it out soon enough.)

**Comment by eseidel** [comment]
I see, I had assumed that I could just set the stack pointer to the simulator's stack pointer and the real cpu would be happy, but maybe that's not a valid assumption. :)

**Comment by eseidel** [comment]
Doesn't work yet, but my attempts from today: https://github.com/dart-lang/sdk/compare/main...shorebirdtech:dart-sdk:simulator_ffi

**Comment by eseidel** [comment]
My assembly didn't end up quite right, but it seems close. Will hack more either this weekend or monday.

**Comment by eseidel** [comment]
I made it work. Code would need a lot more work to be landable (including supporting cross-arch calls), but is enough to unblock the next phase of my investigation: https://github.com/shorebirdtech/dart-sdk/tree/simulator_ffi. Thanks for the help!

**Comment by dcharkes** [comment]
Awesome! 🎉 

> Code would need a lot more work to be landable (including supporting cross-arch calls)

Cross-arch would also mean the stack switching trick doesn't work unfortunately.

Cool keep us posted with the cool stuff! 😎

**Comment by eseidel** [comment]
For those following along at home, my hacks from before work... but now I'm hitting an issue when calling from Dart to C++ and back to Dart and hitting an exception, debugging on:
https://github.com/shorebirdtech/shorebird/issues/654#issuecomment-1604954824

**Comment by eseidel** [comment]
BTW I have this working now for simarm64.  Happy to work with you to upstream.  Right now the diff is in a private repo but happy to bsd license and send it your way when you're interested.  It only works for arm64 atm, and thus would require modifications to support other simulator architectures.

**Comment by dcharkes** [comment]
I don't think we have the bandwidth to add support for other architectures at this point. But if we ever get to it, it would be useful to have your work as a starting point for sure! You could spin up a PR so that we get a copy in Gerrit for future reference.

**Comment by eseidel** [comment]
Happy to.  If you're willing to accept just ARM64 we will eventually post it as a PR (in the middle of launching iOS support atm), since I have little interest in carrying it as a diff.  Thanks!

**Comment by mkustermann** [comment]
Maybe some side note:

We only keep the simulators around for testing purposes (to avoid the need to test on real HW). Though with qemu user mode one can get a pretty good experience with normal arm, since one can use host gdb for debugging of a process running under qemu-arm. So the simulators are somewhat in maintenance mode / on life support.

Supporting FFI for simulators may not even be possible in it's full extent, as the simulated code is targeting ARM abi whereas the native code it interacts with has different abi. Simple conversions between the 2 ABIs can be done of course - but if simulated & host architecture disagree on e.g. bits an `long` has / struct layout it would fall apart. It does work of course for shorebird's use case where host and target architecture are the same.

**Comment by a-siva** [comment]
I have created a sub issue  (https://github.com/dart-lang/sdk/issues/60204) which is a special case of supporting dart:ffi on simarm64 when the host and target architectures are the same.

**Comment by modulovalue** [comment]
We could replace some legacy native bindings with leaf ffi native bindings to get noticeable performance improvements. We can't land those changes because the simulated architectures would also need to support ffi, and they don't.

These are blocked by this issue:
 - https://dart-review.googlesource.com/c/sdk/+/505140
 - https://dart-review.googlesource.com/c/sdk/+/505761

**Comment by dcharkes** [comment]
@mraleph recently suggested we could implement only a subset of `dart:ffi` on the simulator: the subset needed for our own built-in implementations.

**Comment by shikharish** [comment]
@dcharkes I will take this up.
